In [341]:
#import packages
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
# read in file and create dataframe
df = pd.read_csv('advance_tape.csv')

Please analyze the data and provide us insights about our advance originations and
performance data. Produce a light report on your findings.
Here are examples of questions we are looking for within the data:
1. Default rates by loan amount
2. Advance originations by different segmentations

## Exploratoy Data Analysis

In [4]:
# look at first 10 rows of data
df.head(10)

user_id         id origination_date  amount  banks location_state  \
0  00005d618  dab032b8f       2018-09-18     170   Citi             NY   
1  00005d618  ef2879a5d       2018-04-27     130  Wells             TX   
2  0000759ff  8739db27d       2018-04-26     170  Chime             FL   
3  0000759ff  4d27ee400       2018-11-02     140  Wells             GA   
4  0000759ff  18e3eb77d       2018-09-01     170  Chase             CA   
5  000092b1c  0ff223e62       2018-07-18     150  Wells             CA   
6  000092b1c  d4e023bab       2018-11-27     120  Chime             FL   
7  000092b1c  375b599a5       2018-08-18     180  Chase             NY   
8  000092b1c  cdaa1c0bc       2018-06-03     200  Wells             FL   
9  000092b1c  cfa53eb7f       2018-12-16     240  Chase             FL   

   default_flag pay_frequency  
0             1       Monthly  
1             0     Irregular  
2             0     Irregular  
3             0     Irregular  
4             0       Monthly  
5             0     Irregular  
6             0        Weekly  
7             0        Weekly  
8             1     Irregular  
9             0       Monthly

### Features:
● User_id - Identifier for users
● Id - Unique identifier for advances
● Origination Date- Date of Advance Origination
● Amount - Origination Amount
● Location_State - State of Origination of Advance
● Banks - Bank of Borrower’s Depository Account
● Pay Frequency - Frequency of borrowers paycheck
● Default Flag - Advance has defaulted (Not repaid advance)

In [5]:
df.shape

(302866, 8)

#### There are 302,866 unique advances and 8 features in this dataset.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 302866 entries, 0 to 302865
Data columns (total 8 columns):
user_id             302866 non-null object
id                  302866 non-null object
origination_date    302866 non-null object
amount              302866 non-null int64
banks               302866 non-null object
location_state      302866 non-null object
default_flag        302866 non-null int64
pay_frequency       302866 non-null object
dtypes: int64(2), object(6)
memory usage: 18.5+ MB


In [7]:
df.describe()

amount   default_flag
count  302866.000000  302866.000000
mean      160.166212       0.179142
std        48.944875       0.383472
min        80.000000       0.000000
25%       120.000000       0.000000
50%       160.000000       0.000000
75%       200.000000       0.000000
max       240.000000       1.000000

In [8]:
# check for any null values
df.isnull().sum()

user_id             0
id                  0
origination_date    0
amount              0
banks               0
location_state      0
default_flag        0
pay_frequency       0
dtype: int64

#### Data appears to be clean with no Null values that need to be handled.

In [9]:
# convert origination_date to datetime
df['origination_date'] = pd.to_datetime(df['origination_date'])

In [314]:
# date range
print(df['origination_date'].min())
print(df['origination_date'].max())

2018-01-01 00:00:00
2018-12-31 00:00:00


In [21]:
# number of defaults and default percentage
print(df['default_flag'].value_counts())
print('Default Percentage: ' + (str(round(df['default_flag'].mean()*100, 2))))

0    248610
1     54256
Name: default_flag, dtype: int64
Default Percentage: 17.91


In [11]:
# Inspecting the unique values found in columns
df.nunique()

user_id              94009
id                  302866
origination_date       365
amount                  17
banks                    5
location_state           5
default_flag             2
pay_frequency            4
dtype: int64

In [159]:
# confirming that the id values are unique
len(df['id'].unique())

302866

In [333]:
# confirming that the user_id values are unique
len(df['user_id'].unique())

94009

#### There are 94,009 individual users while there are 302,866 advances.

In [13]:
# unique bank values
df['amount'].unique()

array([170, 130, 140, 150, 120, 180, 200, 240, 220, 110, 210, 230,  80,
       160, 190,  90, 100])

In [41]:
# minimum and maximum loan amounts
print(df['amount'].min())
print(df['amount'].max())

80
240


In [336]:
# total number of advances per loan amount
df.groupby(["amount"])['id'].count()

amount
80     17568
90     17736
100    18007
110    17501
120    17826
130    17618
140    17851
150    18007
160    17826
170    17818
180    17867
190    17896
200    17819
210    18001
220    17857
230    17863
240    17805
Name: id, dtype: int64

In [351]:
# total number of defaults per loan amount
df.groupby(["amount", 'default_flag'])['id'].count()

amount  default_flag
80      0               14390
        1                3178
90      0               14596
        1                3140
100     0               14758
        1                3249
110     0               14387
        1                3114
120     0               14547
        1                3279
130     0               14487
        1                3131
140     0               14568
        1                3283
150     0               14767
        1                3240
160     0               14637
        1                3189
170     0               14583
        1                3235
180     0               14679
        1                3188
190     0               14797
        1                3099
200     0               14710
        1                3109
210     0               14794
        1                3207
220     0               14661
        1                3196
230     0               14545
        1                3318
240     0          

In [14]:
# unique bank values
df['banks'].unique()

array(['Citi', 'Wells', 'Chime', 'Chase', 'Goliath'], dtype=object)

In [347]:
# total number of advances per banks
df.groupby(["banks"])['id'].count()

banks
Chase      76121
Chime      30012
Citi       45515
Goliath    90942
Wells      60276
Name: id, dtype: int64

In [354]:
# total number of defaults per bank
df.groupby(["banks", 'default_flag'])['id'].count()

banks    default_flag
Chase    0               62551
         1               13570
Chime    0               24578
         1                5434
Citi     0               37356
         1                8159
Goliath  0               74575
         1               16367
Wells    0               49550
         1               10726
Name: id, dtype: int64

In [15]:
# unique state values
df['location_state'].unique()

array(['NY', 'TX', 'FL', 'GA', 'CA'], dtype=object)

In [346]:
# total number of advances per state
df.groupby(["location_state"])['id'].count()

location_state
CA    76116
FL    60454
GA    30189
NY    45301
TX    90806
Name: id, dtype: int64

In [353]:
# total number of defaults per state
df.groupby(["location_state", 'default_flag'])['id'].count()

location_state  default_flag
CA              0               63487
                1               12629
FL              0               50385
                1               10069
GA              0               25135
                1                5054
NY              0               33876
                1               11425
TX              0               75727
                1               15079
Name: id, dtype: int64

In [16]:
# unique pay frequency values
df['pay_frequency'].unique()

array(['Monthly', 'Irregular', 'Weekly', 'Biweekly'], dtype=object)

In [348]:
# total number of advances per pay frequency
df.groupby(["pay_frequency"])['id'].count()

pay_frequency
Biweekly     75988
Irregular    75857
Monthly      60317
Weekly       90704
Name: id, dtype: int64

In [355]:
# total number of defaults per pay frequency
df.groupby(["pay_frequency", 'default_flag'])['id'].count()

pay_frequency  default_flag
Biweekly       0               62311
               1               13677
Irregular      0               62260
               1               13597
Monthly        0               49613
               1               10704
Weekly         0               74426
               1               16278
Name: id, dtype: int64

## Data Analysis

### Default Analysis

In [359]:
# create function to calculate default ratios 
def by_bank(bank, df):
    ratio_to_total_adv = []
    for b in df[bank].unique():
        ratio_to_total_adv.append(str(b) + ' Ratio Of Defaults To Total ' + b + ' Advances: ' + \
                                  (str(round((len(df[(df['banks'] == b) & \
                                          (df['default_flag'] == 1)])/(df['banks'] == b).sum()), 3))))
    return ratio_to_total_adv


# create function to calculate default amount ratios 
def by_amount(amount, df):
    ratio_to_total_adv = []
    for a in df[amount].unique():
        ratio_to_total_adv.append(str(a) + ' Ratio Of Defaults To Total ' + str(a) + ' Advances: ' + \
                                  (str(round((len(df[(df['amount'] == a) & \
                                          (df['default_flag'] == 1)])/(df['amount'] == a).sum()), 3))))
    return ratio_to_total_adv
 


# create function to calculate default pay frequency ratios 
def pay_freq(frequency, df):
    ratio_to_total_adv = []
    for f in df[frequency].unique():
        ratio_to_total_adv.append(f + ' Ratio Of Defaults To Total ' + f + ' Advances: ' + \
                                  (str(round((len(df[(df['pay_frequency'] == f) & \
                                          (df['default_flag'] == 1)])/(df['pay_frequency'] == f).sum()), 3))))
    return ratio_to_total_adv


# create function to calculate default state ratios  
def by_state(state, df):
    ratio_to_total_adv = []
    for s in df[state].unique():
        ratio_to_total_adv.append(str(s) + ' Ratio Of Defaults To Total ' + s + ' Advances: ' + \
                                  (str(round((len(df[(df['location_state'] == s) & \
                                          (df['default_flag'] == 1)])/(df['location_state'] == s).sum()), 3))))
    return ratio_to_total_adv

In [360]:
by_bank('banks', df)

['Citi Ratio Of Defaults To Total Citi Advances: 0.179',
 'Wells Ratio Of Defaults To Total Wells Advances: 0.178',
 'Chime Ratio Of Defaults To Total Chime Advances: 0.181',
 'Chase Ratio Of Defaults To Total Chase Advances: 0.178',
 'Goliath Ratio Of Defaults To Total Goliath Advances: 0.18']

In [361]:
by_amount('amount', df)

['170 Ratio Of Defaults To Total 170 Advances: 0.182',
 '130 Ratio Of Defaults To Total 130 Advances: 0.178',
 '140 Ratio Of Defaults To Total 140 Advances: 0.184',
 '150 Ratio Of Defaults To Total 150 Advances: 0.18',
 '120 Ratio Of Defaults To Total 120 Advances: 0.184',
 '180 Ratio Of Defaults To Total 180 Advances: 0.178',
 '200 Ratio Of Defaults To Total 200 Advances: 0.174',
 '240 Ratio Of Defaults To Total 240 Advances: 0.174',
 '220 Ratio Of Defaults To Total 220 Advances: 0.179',
 '110 Ratio Of Defaults To Total 110 Advances: 0.178',
 '210 Ratio Of Defaults To Total 210 Advances: 0.178',
 '230 Ratio Of Defaults To Total 230 Advances: 0.186',
 '80 Ratio Of Defaults To Total 80 Advances: 0.181',
 '160 Ratio Of Defaults To Total 160 Advances: 0.179',
 '190 Ratio Of Defaults To Total 190 Advances: 0.173',
 '90 Ratio Of Defaults To Total 90 Advances: 0.177',
 '100 Ratio Of Defaults To Total 100 Advances: 0.18']

In [362]:
pay_freq('pay_frequency', df)

['Monthly Ratio Of Defaults To Total Monthly Advances: 0.177',
 'Irregular Ratio Of Defaults To Total Irregular Advances: 0.179',
 'Weekly Ratio Of Defaults To Total Weekly Advances: 0.179',
 'Biweekly Ratio Of Defaults To Total Biweekly Advances: 0.18']

In [363]:
by_state('location_state', df)

['NY Ratio Of Defaults To Total NY Advances: 0.252',
 'TX Ratio Of Defaults To Total TX Advances: 0.166',
 'FL Ratio Of Defaults To Total FL Advances: 0.167',
 'GA Ratio Of Defaults To Total GA Advances: 0.167',
 'CA Ratio Of Defaults To Total CA Advances: 0.166']

## Visualize Data

#### All visualizations were done in Tableau with labels added in Photoshop.

## Future Steps/Analysis

In [332]:
df.groupby(['user_id', 'id', 'default_flag']).sum().head(50)

amount
user_id   id        default_flag        
00005d618 dab032b8f 1                170
          ef2879a5d 0                130
0000759ff 18e3eb77d 0                170
          4d27ee400 0                140
          8739db27d 0                170
000092b1c 0ff223e62 0                150
          375b599a5 0                180
          910a8c009 0                130
          cdaa1c0bc 1                200
          cfa53eb7f 0                240
          d4e023bab 0                120
0000b8c5a 52003d836 0                110
          5a06d6a43 0                210
          940849ca5 1                230
          aa2bf1c6f 0                130
          bbf0f24da 0                220
          f8f3af8c0 0                110
0000f32f8 b87eb98bd 0                 80
000121080 31878f2e5 0                210
          a7fc9f01e 1                160
0001f48f6 0cbf8e044 1                210
          9ea480983 0                 80
0002caace 79f626ddb 0                170
00033bbff 4d3237157 0                190
          6b81a0212 0                210
          d2ab9ba63 0                210
0004677d7 14ede528d 0                180
0005b3f5c 0fb81599b 0                210
          78d3dbb93 0                 80
0005cb0a0 250566872 0                130
0005d5bf7 854d5d7f1 0                180
          b7297671b 0                190
000709140 236798995 0                170
          2e0d48735 0                190
          7b50194c4 0                170
          810460144 0                120
          96f83cef5 0                240
          a6eaa742e 1                140
000797203 22d59c69a 0                130
          40f2f2398 0                110
          52b3c7bdb 0                 80
          5fba523d0 1                170
          6a3bdce56 0                210
          c5df0f4a1 0                190
0007d9431 37b2deacc 0                190
          38ba9f5d3 1                 90
          3aeb89e8e 0                140
          98d1a6ac3 0                150
          a1eb93688 0                200
          c0c805d08 1                130

In [311]:
df.groupby(["user_id", "default_flag"])['id'].count()

user_id    default_flag
00005d618  0               1
           1               1
0000759ff  0               3
000092b1c  0               5
           1               1
                          ..
fffe635af  0               1
ffff216cc  0               4
           1               1
ffffa14fc  0               8
           1               1
Name: id, Length: 127011, dtype: int64

In [324]:
df[df['user_id'] == '0007d9431']

user_id         id origination_date  amount    banks location_state  \
44  0007d9431  3aeb89e8e       2018-09-26     140    Wells             GA   
45  0007d9431  a1eb93688       2018-09-26     200  Goliath             GA   
46  0007d9431  c0c805d08       2018-10-28     130  Goliath             TX   
47  0007d9431  38ba9f5d3       2018-05-12      90  Goliath             TX   
48  0007d9431  98d1a6ac3       2018-12-06     150  Goliath             CA   
49  0007d9431  c9b4610b2       2018-05-30     110  Goliath             TX   
50  0007d9431  37b2deacc       2018-11-10     190  Goliath             CA   

    default_flag pay_frequency  
44             0     Irregular  
45             0       Monthly  
46             1     Irregular  
47             1     Irregular  
48             0       Monthly  
49             1     Irregular  
50             0     Irregular

In [317]:
df[df['user_id'] == '00005d618']

user_id         id origination_date  amount  banks location_state  \
0  00005d618  dab032b8f       2018-09-18     170   Citi             NY   
1  00005d618  ef2879a5d       2018-04-27     130  Wells             TX   

   default_flag pay_frequency  
0             1       Monthly  
1             0     Irregular

In [27]:
df.groupby(["location_state", "default_flag"]).count()

user_id     id  origination_date  amount  banks  \
location_state default_flag                                                    
CA             0               63487  63487             63487   63487  63487   
               1               12629  12629             12629   12629  12629   
FL             0               50385  50385             50385   50385  50385   
               1               10069  10069             10069   10069  10069   
GA             0               25135  25135             25135   25135  25135   
               1                5054   5054              5054    5054   5054   
NY             0               33876  33876             33876   33876  33876   
               1               11425  11425             11425   11425  11425   
TX             0               75727  75727             75727   75727  75727   
               1               15079  15079             15079   15079  15079   

                             pay_frequency  
location_state default_flag                 
CA             0                     63487  
               1                     12629  
FL             0                     50385  
               1                     10069  
GA             0                     25135  
               1                      5054  
NY             0                     33876  
               1                     11425  
TX             0                     75727  
               1                     15079

In [49]:
df.groupby(["pay_frequency", "default_flag"]).count()

user_id     id  origination_date  amount  banks  \
pay_frequency default_flag                                                    
Biweekly      0               62311  62311             62311   62311  62311   
              1               13677  13677             13677   13677  13677   
Irregular     0               62260  62260             62260   62260  62260   
              1               13597  13597             13597   13597  13597   
Monthly       0               49613  49613             49613   49613  49613   
              1               10704  10704             10704   10704  10704   
Weekly        0               74426  74426             74426   74426  74426   
              1               16278  16278             16278   16278  16278   

                            location_state  
pay_frequency default_flag                  
Biweekly      0                      62311  
              1                      13677  
Irregular     0                      62260  
              1                      13597  
Monthly       0                      49613  
              1                      10704  
Weekly        0                      74426  
              1                      16278